api/models/metadata_test.go

```diff
@@ -0,0 +1,211 @@
+package models
+
+import (
+	"testing"
+	"encoding/json"
+	"reflect"
+	"strings"
+	"fmt"
+)
+
+type testObj struct {
+	Md Metadata `json:"metadata,omitempty"`
+}
+
+type myJson struct {
+	Foo string `json:"foo,omitempty"`
+	Bar string `json:"bar,omitempty"`
+}
+
+func (m Metadata) withRawKey(key string, val string) Metadata {
+	newMd := make(Metadata)
+	for k, v := range m {
+		newMd[k] = v
+	}
+
+	newMd[key] = &metadataValue{val: []byte(val)}
+	return newMd
+}
+
+func TestMetadataEqual(t *testing.T) {
+	mdWithVal, _ := EmptyMetadata().With("foo", "Bar")
+
+	tcs := []struct {
+		a      Metadata
+		b      Metadata
+		equals bool
+	}{
+
+		{EmptyMetadata(), EmptyMetadata(), true},
+		{mdWithVal, EmptyMetadata(), false},
+		{mdWithVal, mdWithVal, true},
+		{mdWithVal.Without("foo"), EmptyMetadata(), true},
+	}
+
+	for _, tc := range tcs {
+		if tc.a.Equals(tc.b) != tc.equals {
+			t.Errorf("Metadata equality mismatch - expecting (%v == %v) = %v", tc.b, tc.a, tc.equals)
+		}
+		if tc.b.Equals(tc.a) != tc.equals {
+			t.Errorf("Metadata reflexive equality mismatch - expecting (%v == %v) = %v", tc.b, tc.a, tc.equals)
+		}
+	}
+
+}
+
+var mdCases = []struct {
+	val       *testObj
+	valString string
+}{
+	{val: &testObj{Md: EmptyMetadata()}, valString: "{}"},
+	{val: &testObj{Md: EmptyMetadata().withRawKey("stringval", "\"bar\"")}, valString: "{\"metadata\":{\"stringval\":\"bar\"}}"},
+	{val: &testObj{Md: EmptyMetadata().withRawKey("intval", "1001")}, valString: "{\"metadata\":{\"intval\":1001}}"},
+	{val: &testObj{Md: EmptyMetadata().withRawKey("floatval", "3.141")}, valString: "{\"metadata\":{\"floatval\":3.141}}"},
+	{val: &testObj{Md: EmptyMetadata().withRawKey("objval", "{\"foo\":\"fooval\",\"bar\":\"barval\"}")}, valString: "{\"metadata\":{\"objval\":{\"foo\":\"fooval\",\"bar\":\"barval\"}}}"},
```

Comment:
should 
```
{val: &testObj{Md: EmptyMetadata().withRawKey("objval", "{\"foo\":\"fooval\",\"bar\":{\"barbar\":\"barbarval\"}")}, valString: "{\"metadata\":{\"objval\":{\"foo\":\"fooval\",\"bar\":{\"barbar\":\"barbarval\"}}}}"}
```
work?

Reply:
I think so   - what looks wrong?

Reply:
ah got it...

Reply:
You are missing a bracket, but yes it does work

Reply:
metadata values are "any json value except the empty string or null"  - including nested options